sdk/typescript/src/rpc/websocket-client.ts

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { isSubscriptionEvent } from '../types/index.guard';
+import {
+  SuiEventFilter,
+  SuiEventEnvelope,
+  SubscriptionId,
+} from '../types';
+import { Client as WsRpcClient} from 'rpc-websockets';
+
+
+export const getWebsocketUrl = (httpUrl: string, port?: number): string => {
+  const url = new URL(httpUrl);
+  url.protocol = url.protocol.replace('http', 'ws');
+  url.port = (port ?? 9001).toString();
+  return url.toString();
+};
+
+enum ConnectionState {
+  NotConnected,
+  Connecting,
+  Connected
+}
+
+type JsonRpcMethodMessage<T> = {
+  jsonrpc: '2.0',
+  method: string,
+  params: T
+}
+
+type FilterSubHandler = {
+  id: SubscriptionId,
+  onMessage: (event: SuiEventEnvelope) => void,
+  filter: SuiEventFilter
+};
+
+type SubscriptionData = {
+  filter: SuiEventFilter,
+  onMessage: (event: SuiEventEnvelope) => void
+}
+
+type MinimumSubscriptionMessage = {
+  subscription: SubscriptionId,
+  result: object
+}
+
+const isMinimumSubscriptionMessage = (msg: any): msg is MinimumSubscriptionMessage =>
+  msg
+  && ('subscription' in msg && typeof msg['subscription'] === 'number')
+  && ('result' in msg && typeof msg['result'] === 'object')
+
+export type WebsocketClientOptions = {
+  connectTimeout: number,
+  callTimeout: number,
+  reconnectInterval: number,
+  maxReconnects: number
+}
+
+export const DEFAULT_CLIENT_OPTIONS: WebsocketClientOptions = {
+  connectTimeout: 15000,
+  callTimeout: 30000,
+  reconnectInterval: 3000,
+  maxReconnects: 5
+}
+
+const SUBSCRIBE_EVENT_METHOD = 'sui_subscribeEvent';
+const UNSUBSCRIBE_EVENT_METHOD = 'sui_unsubscribeEvent';
+
+export class WebsocketClient {
+  protected rpcClient: WsRpcClient;
+  protected connectionState: ConnectionState = ConnectionState.NotConnected;
+  protected connectionTimeout: number | null = null;
+  protected isSetup: boolean = false;
+
+  protected eventSubscriptions: Map<SubscriptionId, SubscriptionData> = new Map();
+
+  constructor(
+    public endpoint: string,
+    public skipValidation: boolean,
+    public options: WebsocketClientOptions = DEFAULT_CLIENT_OPTIONS
+  ) {
+    if (this.endpoint.startsWith('http'))
+      this.endpoint = getWebsocketUrl(this.endpoint);
+
+    this.rpcClient = new WsRpcClient(this.endpoint, {
+      reconnect_interval: this.options.reconnectInterval,
+      max_reconnects: this.options.maxReconnects,
+      autoconnect: false
```

Comment:
why is this hardcoded to false?

Reply:
discussed on call: to prevent automatically connecting, which is the default